openplacereviews-core/src/main/java/org/openplacereviews/osm/service/MergePlaceBot.java

```diff
@@ -211,6 +213,141 @@ protected void mergePlaces(OprMapCollectionApiResult res, MergeInfo info) {
+				placesToMerge = filterByDistance(placesToMerge, deleted);
+				placesToMerge = filterByOsmId(placesToMerge, deleted);
+				if (placesToMerge.isEmpty() || !foundPlaceWithSameName(placesToMerge, deleted)) {
+					return permanentlyClosePlace(deleted, edited);
+				}
+			}
+		}
+		return false;
+	}
+
+	private boolean permanentlyClosePlace(OpObject oldObj, List<OpObject> edited) {
+		OpObject newObj = new OpObject(oldObj, false);
+		newObj.setFieldByExpr(F_DELETED_PLACE, Instant.now().toString());
+		addObjToOperation(oldObj, newObj, null, edited);
+		return true;
+	}
+
+	private List<Feature> filterByDistance(List<Feature> placesToMerge, OpObject deleted) {
+		LatLon latLonDeleted = getLatLon(deleted);
+		List<Feature> res = new ArrayList<>();
+		if (latLonDeleted != null) {
+			for (Feature feature : placesToMerge) {
+				Point featurePoint = (Point) feature.geometry();
+				if (OsmMapUtils.getDistance(latLonDeleted.getLatitude(), latLonDeleted.getLongitude(), featurePoint.lat(), featurePoint.lon()) < 150) {
+					res.add(feature);
+				}
+			}
+		}
+		return res;
+	}
+
+	private List<Feature> filterByOsmId(List<Feature> placesToMerge, OpObject deleted) {
+		String osmIdDeleted = getOsmId(deleted);
+		List<Feature> res = new ArrayList<>();
+		if (osmIdDeleted != null) {
+			for (Feature feature : placesToMerge) {
+				if (!feature.properties().get(OSM_ID).toString().equals(osmIdDeleted)) {
+					res.add(feature);
+				}
+			}
+		}
+		return res;
+	}
+
+	private boolean foundPlaceWithSameName(List<Feature> placesToMerge, OpObject deleted) {
+		List<Map<String, String>> placesToMergeTags = new ArrayList<>();
+		Map<String, String> oldOsmTags = getMainOsmTags(deleted);
+		if (oldOsmTags != null) {
+			List<OpObject> objToMerge = new ArrayList<>();
+			for (Feature feature : placesToMerge) {
+				objToMerge.add(getCurrentObject(feature));
+			}
+			for (int i = 0; i < placesToMerge.size(); i++) {
+				placesToMergeTags.add(getMainOsmTags(objToMerge.get(i)));
+			}
+			for (MatchType mt : EnumSet.allOf(MatchType.class)) {
+				int matched = -1;
+				for (int i = 0; i < placesToMerge.size(); i++) {
+					if (match(mt, oldOsmTags, placesToMergeTags.get(i))) {
+						if (matched >= 0) {
+							if (mt.allow2PlacesMerge) {
+								matched = i;
+							} else {
+								return false;
+							}
+						} else {
+							matched = i;
+						}
+					}
+				}
+				if (matched >= 0) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	private List<Feature> getNoDeletedPlaces(List<Feature> features) {
+		List<Feature> res = new ArrayList<>();
+		for (Feature feature : features) {
+			if (!feature.properties().containsKey(F_DELETED_PLACE)) {
+				Map<String, Object> osm = getMainOsmFromList(getCurrentObject(feature));
+				if (osm != null && !osm.containsKey(F_DELETED_OSM)) {
+					res.add(feature);
+				}
+			}
+		}
+		return res;
+	}
+
+	private LocalDate transformStringDateToLocaleDate(String date) {
```

Comment:
doesn't need to be a online line method as it looks. Data parsing is not inlined with formatting. There should be use of constant that was specified during write